### ISSUE TITLE
Interpreting probabilities in BERTopic vs. LDA results

### ISSUE THREAD

**Issue by rja122277** [MaartenGr/BERTopic#763]
Hi, 

I'm wondering what the differences between "probabilities" calculated from the BERTopic model and the LDA model are. (or do they mean the same thing?)

I'm a beginner in this field and what I've learned so far is that BERTopic and LDA use different approaches to extract topics - LDA is a generative statistical model and BERTopic uses an embedding approach. But both of them produce the "topic probabilities", and this term confuses me. Because some say that the BERTopic is not based on statistical learning (and that's how it performs better than LDA, especially when analyzing the short-length text), it is not intuitive for me to connect the term "probabilities" with the BERTopic's results.

The BERTopic tutorial says "probabilities: The probability of the assigned topic per document. These are only calculated if a HDBSCAN model is used for the clustering step. When calculate_probabilities=True, then it is the probabilities of all topics per document." Is it a correct way to interpret that the probability of Topic A in a document means the likelihood that a word from Topic A appears in the given document? Then how does this interpretation differ from the probabilities resulting from LDA analysis? 

Many thanks in advance!

**Comment by MaartenGr** [comment]
In short, the probabilities are a result of the clustering with HDBSCAN. They are the probabilities that a document belongs to a certain cluster. You can find more about that [here](https://hdbscan.readthedocs.io/en/latest/soft_clustering.html). Similarly, [this guide](https://maartengr.github.io/BERTopic/algorithm/algorithm.html) to BERTopic might also help you get a more intuitive understanding of the algorithm. LDA and BERTopic are difficult to compare as they are quite different algorithms and approach the topic modeling task entirely different. With LDA, documents contain a mixture of topics whereas BERTopic reduces the topic modeling approach to a clustering task. As a result, the probabilities can mean something quite different but can be used similarly.

**Comment by rja122277** [comment]
Thanks a lot! I read those references and now understand more clearly. So the probabilities are produced during the clustering stage. The probs mean the likelihood that a document (represented as a vector with a reduced dimension) is included in a cluster (rather than other clusters). After this process, BERTopic creates a bag-of-words representation by combining all documents in each topic and extracting the most frequently appearing words on a cluster-level. 

One clarifying question is - then does this BERTopic assume each document will have one consistent meaning or be categorized into one consistent cluster? This is because they cluster (categorize) each document (made of a vector of different dimensions) into one group.

And then, is there any way to calculate to what extent Document  A is similar/consistent with Document  B in terms of different topics (clusters) the model produces? I think this question loosely relates to the discussion above - (1) We can compare different documents by using their vectors, like calculating cosine similarity or others. (2) But this vector is supposed to lie into only one cluster. (3) So I'm not sure if it is possible to compare how much Document  A is similar with Document B in terms of entire clusters (topics) created by the model . 

I don't know if this makes sense at all. Please let me know if I take something wrong here!

**Comment by MaartenGr** [comment]
> One clarifying question is - then does this BERTopic assume each document will have one consistent meaning or be categorized into one consistent cluster? 

Yes, since we are essentially clustering our documents it is assumed that each document belongs to a single document. However, we then extract topic representations from those clusters and as a result, we essentially extract the *dominant* topic in a document. This does not mean that BERTopic assumes each document will have one consistent meaning but merely that it only extracts its main topic. 

Depending on your dataset, that can definitely be a disadvantage. What you can do instead is split up the documents into sentences and let BERTopic cluster those instead. 

> This is because they cluster (categorize) each document (made of a vector of different dimensions) into one group.

Do note that all vectors in BERTopic generated through the language model have the same dimensions. 

> And then, is there any way to calculate to what extent Document A is similar/consistent with Document B in terms of different topics (clusters) the model produces? I think this question loosely relates to the discussion above - (1) We can compare different documents by using their vectors, like calculating cosine similarity or others. (2) But this vector is supposed to lie into only one cluster. (3) So I'm not sure if it is possible to compare how much Document A is similar with Document B in terms of entire clusters (topics) created by the model .

You can do what you suggested here since all vectors generated by the language model have the same dimensional space and therefore can be compared with one another. You can even apply the c-TF-IDF on the documents and compare its output with the c-TF-IDF representation of its topics. 

You can also use the `calculate_probabilities=True` parameter to generate a topic-document probability matrix that essentially tells you the probability of a document belonging to all topics. This is possible since our clustering algorithm, HDBSCAN, has the option to perform soft-clustering after doing its initial clustering task. 

I don't know if this makes sense at all. Please let me know if I take something wrong here!

**Comment by rja122277** [comment]
To clarify, the "probabilities" I mentioned in my earlier threads indicate (a topic-document) probability resulting from the `calculate_probabilities=True` parameter. Your explanation broadens my understanding of using this prob value. Many thanks! 

> You can do what you suggested here since all vectors generated by the language model have the same dimensional space and therefore can be compared with one another. You can even apply the c-TF-IDF on the documents and compare its output with the c-TF-IDF representation of its topics.
To your points on c-TF-IDF vectors, how can I produce and store the c-TF-IDF of each document? I need this information to calculate the cosine similarity of two documents' vectors and measure their topic similarity. I assume that I need to produce the (c-TF-IDF) vector of each document first and then put them in the cosine similarity equation. Does it sound like the right way to do it? And if I use a topic-document prob matrix instead of a c-TF-IDF, then what measures can I make with the matrix data to compare the different documents? I want to understand the usage of these different products in BERTopic model so that I can use the model and construct measures in my research.

(To give a background, my sample structure is a firm-year-document panel data. Multiple documents (about 5sentences-long) are available for each firm-year, and each firm has multiple years. For example, Apple has 5 documents in 2010, 6 documents in 2011, and 5 documents in 2012. I want to see how much the topics of documents change/stay consistent over time for each firm. So I'm thinking of getting the vectors of each document, averaging them into a firm-year, and then calculating the cosine similarity of two consecutive years' vectors.)

**Comment by MaartenGr** [comment]
> To your points on c-TF-IDF vectors, how can I produce and store the c-TF-IDF of each document? I assume that I need to produce the (c-TF-IDF) vector of each document first and then put them in the cosine similarity equation if I want to measure the topic similarity of two documents. Does it make sense at all?

**Comment by MaartenGr** [comment]
> To your points on c-TF-IDF vectors, how can I produce and store the c-TF-IDF of each document? I assume that I need to produce the (c-TF-IDF) vector of each document first and then put them in the cosine similarity equation if I want to measure the topic similarity of two documents. Does it make sense at all?

In other to create the c-TF-IDF vectors for each document, you can perform the following:

```python
bag_of_words = topic_model.vectorizer_model.transform(documents)
c_tf_idf_per_document = topic_model.ctfidf_model.transform(bag_of_words)
```

You can then compare `c_tf_idf_per_document`  with `topic_model.c_tf_idf_`. 

> On the other, if I use a topic-document prob matrix instead of a c-TF-IDF, then what measures can I make with the matrix data to compare the different documents? I want to understand the usage of these different products in BERTopic model so that I can use the model and construct measures in my research.

The probabilities given in the topic-document probability matrix already indicate the likelihood a topic is found within a document. You can then take the top 5 topics per document as long as they pass a certain user-defined threshold.

**Comment by rja122277** [comment]
Thank you! This is really helpful. I really appreciate it. But I got an error (ValueError: dimension mismatch) when I tried to create the `c_tf_idf_per_document` by reloading the saved bertopic models. 

Specifically, the codes reported the error on one model but worked well on the other model. I analyzed two models - In the first model, I used `ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)`. In the second model, `I used ctfidf_model = ClassTfidfTransformer()` . All other parameters and documents in use were same. I analyzed these two different models and saved them individually. Then I reloaded them to create `c_tf_idf_per_document` and calculate cosine similarity between documents. But as mentioned above, the codes didn't work for the first model. The first line (creating `bag_of_words`) seems to work, but the second line doesn't. I attached the screenshot of the error for your information. Thanks for your time in advance!


![image](https://user-images.githubusercontent.com/113876288/200675717-6f8ddab9-b2a9-4f8b-ae06-81c8135c1a50.png)

**Comment by rja122277** [comment]
Plus, I found that `ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)` vs. `ctfidf_model = ClassTfidfTransformer() ` somtimes generate very different numbers of topics even when I used the same parameters (e.g., n_neighbors, min_cluster, min_sample). With some combination of parameters, both conditions generate similar number of topics. But in the case of other combinations, the numbers of topic they produce are very different like 62 vs.2 or 77 vs. 3. They even don't have patterns like `(reduce_frequent_words=True)` produce greater number of topics. Could you please explain why this is happening? Is this even expected to happen or I'm doing something wrong here?

**Comment by MaartenGr** [comment]
> Thank you! This is really helpful. I really appreciate it. But I got an error (ValueError: dimension mismatch) when I tried to create the c_tf_idf_per_document by reloading the saved bertopic models.

I am not entirely sure what is happening here without a bit more context. Is `data1sent` a list or a string? How did you train your topic model? Could you share all code before coming to this error? 

> Plus, I found that ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True) vs. ctfidf_model = ClassTfidfTransformer() somtimes generate very different numbers of topics even when I used the same parameters (e.g., n_neighbors, min_cluster, min_sample). With some combination of parameters, both conditions generate similar number of topics. But in the case of other combinations, the numbers of topic they produce are very different like 62 vs.2 or 77 vs. 3. They even don't have patterns like (reduce_frequent_words=True) produce greater number of topics. Could you please explain why this is happening? Is this even expected to happen or I'm doing something wrong here?

The underlying dimensionality reduction model in BERTopic is UMAP, which is a stochastic model. This means that the results can differ between runs. You can find a bit more about that [here](https://maartengr.github.io/BERTopic/faq.html#why-are-the-results-not-consistent-between-runs).

**Comment by rja122277** [comment]
First, `data1sent `is a list. Here are two models I ran: 

```
embedding_model = SentenceTransformer('all-mpnet-base-v2')
vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words=stopwords)
umap_model = UMAP(n_neighbors=10)
hdbscan_model172 = HDBSCAN(min_cluster_size=17, min_samples=2,  
                        gen_min_span_tree=True,
                        prediction_data=True)
ctfidf_model = ClassTfidfTransformer()
ctfidf_model2 = ClassTfidfTransformer(reduce_frequent_words=True)

model1= BERTopic(
    embedding_model=embedding_model,
    umap_model=umap_model,
    hdbscan_model=hdbscan_model172,
    vectorizer_model=vectorizer_model,
    ctfidf_model=ctfidf_model,
    language='english',
    calculate_probabilities=True,
    verbose=True
)
topics1, probs1 = model1.fit_transform(data1sent)

model2 = BERTopic(
    embedding_model=embedding_model,
    umap_model=umap_model,
    hdbscan_model=hdbscan_model172,
    vectorizer_model=vectorizer_model,
    ctfidf_model=ctfidf_model2,
    language='english',
    calculate_probabilities=True,
    verbose=True
)
topics2, probs2 = model2.fit_transform(data1sent)
```
I saved model1 and model2. Then I reloaded them to create `c_tf_idf_per_document` but faced the issue "dimension mismatch" above when using the model2. 

With the model2, I got the same "dimension mismatch" error when I explore `hierarchical_topics` and `topics_over_time`. I used `hierarchical_topics = model2.hierarchical_topics(data1sent)` and  `topics_over_time = model2.topics_over_time(data1sent, timestamps, nr_bins=None)`, respectively. Again, I don't have any problem when using the model1. 

Are these three functions (`c_tf_idf_per_document`, `hierarchical_topics`, and `topics_over_time`) not eligible for the model that already reduced the words with `ClassTfidfTransformer(reduce_frequent_words=True)`? In this case, how can I evaluate which model works better for me?

**Comment by MaartenGr** [comment]
You cannot compare 2 topic models based on their c-TF-IDF values if they have different vocabularies which may be as a result of using different c-TF-IDF models. The c-TF-IDF matrix is essentially a topic-term matrix and if the two of them have a different number of features, then you cannot compare them.

**Comment by rja122277** [comment]
Let me clarify the issue here. I'm not trying to compare 2 topic models here. I'm exploring these two different models individually. And I found that I can run some commands with the model1 but not with the model2. 

Meaning, I'm running these three commands:
```
bag_of_words = topic_model.vectorizer_model.transform(documents)
c_tf_idf_per_document = topic_model.ctfidf_model.transform(bag_of_words)
```
`hierarchical_topics = model2.hierarchical_topics(data1sent)
`
`topics_over_time = model2.topics_over_time(data1sent, timestamps, nr_bins=None)
`

With model1, I can run these three commands successfully and explore the model results. However, with model2, I cannot run these commands. For each of these three commands, the system says "dimension mismatch." So I'm wondering if this is because (as you also assumed) I mix up the model1 and model2 by mistake. But this is less likely because I ran them in two separate files. So I'm wondering if there is any systematic reason that I cannot run these three commands when I use the model, which was produced with the condition of `ctfidf_model2 = ClassTfidfTransformer(reduce_frequent_words=True)`. Please let me know if anything is unclear. Thank you!

**Comment by MaartenGr** [comment]
I just tried out a few things and I cannot seem to replicate the issue. The following is working for me:

```python
from bertopic import BERTopic
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer
from bertopic.vectorizers import ClassTfidfTransformer

# Load data
docs = fetch_20newsgroups(subset='all',  remove=('headers', 'footers', 'quotes'))['data']

# Prepare sub-models
vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words="english")
ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=True)

# Train BERTopic
topic_model = BERTopic()
topics, probs = topic_model.fit_transform(docs)

# Save and Load BERTopic model
topic_model.save("my_model")
loaded_model = BERTopic.load("my_model")

# Generate c-TF-IDF per document
bag_of_words = loaded_model.vectorizer_model.transform(docs)
c_tf_idf_per_document = loaded_model.ctfidf_model.transform(bag_of_words)

bag_of_words = topic_model.vectorizer_model.transform(docs)
c_tf_idf_per_document = topic_model.ctfidf_model.transform(bag_of_words)
```

Seeing your error, I get the feeling that somehow two vectorizer models get mixed up. Which version of BERTopic are you using? Also, you mentioned you saved and loaded the model. Could you share your entire pipeline, without leaving any step out? Perhaps something went wrong with saving the model.

**Comment by rja122277** [comment]
Thanks for the answer! First, I'm using BERTopic version 0.12. 

And yes, I have the same thoughts as you. I feel two models were mixed up somewhere in the process. I reran the whole process to see if the problems still appear. Unfortunately, it appeared again. Here are my codes:

```
# Load the data 
df = pd.read_csv(r'C:/Users/Box/0_topic_modeling/output.csv')
data1sent = df.text1sent.values.tolist()

# Prepare sub-models
embedding_model = SentenceTransformer('all-mpnet-base-v2')
vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words=stopwords)
umap_model = UMAP(n_neighbors=10)
hdbscan_model172 = HDBSCAN(min_cluster_size=17, min_samples=2,  
                        gen_min_span_tree=True,
                        prediction_data=True)
ctfidf_model = ClassTfidfTransformer()
ctfidf_model2 = ClassTfidfTransformer(reduce_frequent_words=True)

# Run BERTopic
model = BERTopic(
    embedding_model=embedding_model,
    umap_model=umap_model10,
    hdbscan_model=hdbscan_model172,
    vectorizer_model=vectorizer_model,
    ctfidf_model=ctfidf_model,
    language='english',
    calculate_probabilities=True,
    verbose=True
)
topics, probs= model.fit_transform(data1sent)

model1 = BERTopic(
    embedding_model=embedding_model,
    umap_model=umap_model10,
    hdbscan_model=hdbscan_model172,
    vectorizer_model=vectorizer_model,
    ctfidf_model=ctfidf_model2,
    language='english',
    calculate_probabilities=True,
    verbose=True
)
topics1, probs1 = model1 .fit_transform(data1sent)

# Save and Load the BERTopic model
model.save("C:/Users/Box/0_topic_modeling/test_ctfidf")
model1.save("C:/Users/Box/0_topic_modeling/test_ctfidf_reduce")

my_model = BERTopic.load("C:/Users/Box/0_topic_modeling/test_ctfidf")
topics = my_model.topics_
probs = my_model.probabilities_
embedding_model = my_model.topic_embeddings_

my_model1 = BERTopic.load("C:/Users/Box/0_topic_modeling/test_ctfidf_reduce")
topics1 = my_model1.topics_
probs1 = my_model1.probabilities_

# Generate c-TF_IDF per document
bag_of_words = my_model.vectorizer_model.transform(data1sent)
c_tf_idf_per_document = my_model.ctfidf_model.transform(bag_of_words)

bag_of_words1 = my_model1.vectorizer_model.transform(data1sent)
c_tf_idf_per_document1 = my_model1.ctfidf_model.transform(bag_of_words)
```

But this time, the errors come up with the first model (my_model) not the second one (my_model1). So maybe these two models are mixed up in some steps... Here is another screen capture of the error. 

<img width="379" alt="err" src="https://user-images.githubusercontent.com/113876288/202559576-bbeb7358-9aea-4a15-99e1-c2efa49b1165.PNG">

**Comment by MaartenGr** [comment]
I took some time but I am quite I figured it out. It seems that since you are passing the exact same `vectorizer_model` to both topic models, it gets changed globally since no internal copies are being made. Instead, I would advise making sure that whatever you pass to one model, like the vectorizer, you create a copy beforehand and then pass it to the other model. In other words, do the following:

```python
vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words=stopwords)
vectorizer_model1 = CountVectorizer(ngram_range=(1, 2), stop_words=stopwords)

model = BERTopic(vectorizer_model=vectorizer_model)
model1 = BERTopic(vectorizer_model=vectorizer_model1)
```

> Lastly, another thing came up in this process - I understand that BERTopic produces the different number of topics everytime I run it again. But the number of my topics decreased from https://github.com/MaartenGr/BERTopic/issues/62 to https://github.com/MaartenGr/BERTopic/pull/3 even when I used the exact same submodels. Is this in the range you expect? In this case, how do I decide what the most reliable model is?

It may just be that the `min_cluster_size` that you have set is on the border of finding a sufficient amount of clusters. Reducing this value is likely to help.

**Comment by rja122277** [comment]
This resolves everything. I did some experiments too, and I found that it can work even when I use the same `vectorizer_model = CountVectorizer(ngram_range=(1, 2), stop_words=stopwords)` for both models, if I redefine this command in each cell where I train the new model. (By the way, I found another mistake in my codes above - I didn't update from `bag_of_words ` to `bag_of_words1`.) An over-the-month-long conversation and the links you attached so far (especially the one about soft clustering) have made my research much stronger. Thanks a lot.
